File: src/shared/pointer-movement-settings.ts
export type PointerMovementSizeKey = 'small' | 'medium' | 'large';

export type PointerMovementSettings = {
  scalePercent: number;
};

export const BASE_POINTER_MOVEMENT_PERCENTAGES: Record<PointerMovementSizeKey, number> = {
  small: 4.5,
  medium: 12,
  large: 26
};

export const DEFAULT_POINTER_MOVEMENT_SETTINGS: PointerMovementSettings = {
  scalePercent: 100
};

export const POINTER_MOVEMENT_SCALE_MIN = 25;
export const POINTER_MOVEMENT_SCALE_MAX = 200;
export const POINTER_MOVEMENT_SCALE_STEP = 5;

const pointerMovementSizeKeys: PointerMovementSizeKey[] = ['small', 'medium', 'large'];
const DISPLAY_PERCENTAGE_MIN = 1;
const DISPLAY_PERCENTAGE_MAX = 50;
const DISPLAY_PERCENTAGE_STEP = 0.5;

export function normalizePointerMovementSettings(value: unknown): PointerMovementSettings {
  if (!value || typeof value !== 'object') {
    return { ...DEFAULT_POINTER_MOVEMENT_SETTINGS };
  }

  const candidate = value as Partial<PointerMovementSettings> & {
    percentages?: unknown;
    multipliers?: unknown;
  };

  if ('scalePercent' in candidate) {
    return { scalePercent: normalizeScale(candidate.scalePercent) };
  }

  if (candidate.percentages && typeof candidate.percentages === 'object') {
    return { scalePercent: normalizeScale(scaleFromPercentages(candidate.percentages)) };
  }

  if (candidate.multipliers && typeof candidate.multipliers === 'object') {
    return { scalePercent: normalizeScale(scaleFromLegacyMultipliers(candidate.multipliers)) };
  }

  return { ...DEFAULT_POINTER_MOVEMENT_SETTINGS };
}

export function pointerMovementScalePercentFor(settings: PointerMovementSettings): number {
  return normalizePointerMovementSettings(settings).scalePercent;
}

export function pointerMovementPercentageFor(
  settings: PointerMovementSettings,
  size: PointerMovementSizeKey
): number {
  const scale = pointerMovementScalePercentFor(settings) / 100;
  return normalizeDisplayPercentage(BASE_POINTER_MOVEMENT_PERCENTAGES[size] * scale);
}

export function pointerMovementFractionFor(
  settings: PointerMovementSettings,
  size: PointerMovementSizeKey
): number {
  return pointerMovementPercentageFor(settings, size) / 100;
}

function scaleFromPercentages(value: unknown): number {
  const percentages = value as Partial<Record<PointerMovementSizeKey, unknown>>;
  const scales = pointerMovementSizeKeys
    .map((size) => {
      const percentage = percentages[size];
      if (typeof percentage !== 'number' || !Number.isFinite(percentage)) return null;
      return (normalizeDisplayPercentage(percentage) / BASE_POINTER_MOVEMENT_PERCENTAGES[size]) * 100;
    })
    .filter((scale): scale is number => scale !== null);

  if (scales.length === 0) return DEFAULT_POINTER_MOVEMENT_SETTINGS.scalePercent;
  return scales.reduce((sum, scale) => sum + scale, 0) / scales.length;
}

function scaleFromLegacyMultipliers(value: unknown): number {
  const multipliers = value as Partial<Record<PointerMovementSizeKey, unknown>>;
  const scales = pointerMovementSizeKeys
    .map((size) => {
      const multiplier = multipliers[size];
      return typeof multiplier === 'number' && Number.isFinite(multiplier) ? multiplier : null;
    })
    .filter((scale): scale is number => scale !== null);

  if (scales.length === 0) return DEFAULT_POINTER_MOVEMENT_SETTINGS.scalePercent;
  return scales.reduce((sum, scale) => sum + scale, 0) / scales.length;
}

function normalizeScale(value: unknown): number {
  if (typeof value !== 'number' || !Number.isFinite(value)) {
    return DEFAULT_POINTER_MOVEMENT_SETTINGS.scalePercent;
  }

  return clamp(roundToStep(value, POINTER_MOVEMENT_SCALE_STEP), POINTER_MOVEMENT_SCALE_MIN, POINTER_MOVEMENT_SCALE_MAX);
}

function normalizeDisplayPercentage(value: number): number {
  return clamp(roundToStep(value, DISPLAY_PERCENTAGE_STEP), DISPLAY_PERCENTAGE_MIN, DISPLAY_PERCENTAGE_MAX);
}

function roundToStep(value: number, step: number): number {
  return Math.round(value / step) * step;
}

function clamp(value: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, value));
}

File: src/SwitchifyPc.Core/Settings/PointerMovementSettings.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using SwitchifyPc.Core.Storage;

namespace SwitchifyPc.Core.Settings;

public enum PointerMovementSizeKey
{
    Small,
    Medium,
    Large
}

public sealed record PointerMovementSettings(double ScalePercent);

public static class PointerMovementSettingsModel
{
    public const double PointerMovementScaleMin = 5;
    public const double PointerMovementScaleMax = 225;
    public const double PointerMovementScaleStep = 5;
    public const int BaseMoveDelta = 128;

    private const double DisplayPercentageMin = 1;
    private const double DisplayPercentageMax = 50;
    private const double DisplayPercentageStep = 0.5;

    public static readonly PointerMovementSettings Default = new(100);

    public static readonly IReadOnlyDictionary<PointerMovementSizeKey, double> BasePercentages =
        new Dictionary<PointerMovementSizeKey, double>
        {
            [PointerMovementSizeKey.Small] = 4.5,
            [PointerMovementSizeKey.Medium] = 12,
            [PointerMovementSizeKey.Large] = 26
        };

    public static PointerMovementSettings Normalize(JsonNode? value)
    {
        if (value is not JsonObject candidate)
        {
            return Default;
        }

        if (candidate.TryGetPropertyValue("scalePercent", out JsonNode? scalePercent))
        {
            return new PointerMovementSettings(NormalizeScale(scalePercent));
        }

        if (candidate.TryGetPropertyValue("percentages", out JsonNode? percentages) && percentages is JsonObject percentageObject)
        {
            return new PointerMovementSettings(NormalizeScale(ScaleFromPercentages(percentageObject)));
        }

        if (candidate.TryGetPropertyValue("multipliers", out JsonNode? multipliers) && multipliers is JsonObject multiplierObject)
        {
            return new PointerMovementSettings(NormalizeScale(ScaleFromLegacyMultipliers(multiplierObject)));
        }

        return Default;
    }

    public static PointerMovementSettings Normalize(object? value)
    {
        return Normalize(JsonSerializer.SerializeToNode(value));
    }

    public static double ScalePercentFor(PointerMovementSettings settings)
    {
        return Normalize(settings).ScalePercent;
    }

    public static double PercentageFor(PointerMovementSettings settings, PointerMovementSizeKey size)
    {
        double scale = ScalePercentFor(settings) / 100;
        return NormalizeDisplayPercentage(BasePercentages[size] * scale);
    }

    public static double FractionFor(PointerMovementSettings settings, PointerMovementSizeKey size)
    {
        return PercentageFor(settings, size) / 100;
    }

    public static JsonObject ToJsonObject(PointerMovementSettings settings)
    {
        PointerMovementSettings normalized = Normalize(settings);
        return new JsonObject
        {
            ["scalePercent"] = normalized.ScalePercent
        };
    }

    private static double NormalizeScale(JsonNode? value)
    {
        return TryGetFiniteNumber(value, out double number)
            ? Clamp(RoundToStep(number, PointerMovementScaleStep), PointerMovementScaleMin, PointerMovementScaleMax)
            : Default.ScalePercent;
    }

    public static PointerMovementSettings Normalize(PointerMovementSettings settings)
    {
        return new PointerMovementSettings(Clamp(RoundToStep(settings.ScalePercent, PointerMovementScaleStep), PointerMovementScaleMin, PointerMovementScaleMax));
    }

    private static double ScaleFromPercentages(JsonObject value)
    {
        List<double> scales = [];
        foreach (PointerMovementSizeKey size in Enum.GetValues<PointerMovementSizeKey>())
        {
            if (value.TryGetPropertyValue(JsonKey(size), out JsonNode? percentage) && TryGetFiniteNumber(percentage, out double number))
            {
                scales.Add((NormalizeDisplayPercentage(number) / BasePercentages[size]) * 100);
            }
        }

        return scales.Count == 0 ? Default.ScalePercent : scales.Average();
    }

    private static double ScaleFromLegacyMultipliers(JsonObject value)
    {
        List<double> scales = [];
        foreach (PointerMovementSizeKey size in Enum.GetValues<PointerMovementSizeKey>())
        {
            if (value.TryGetPropertyValue(JsonKey(size), out JsonNode? multiplier) && TryGetFiniteNumber(multiplier, out double number))
            {
                scales.Add(number);
            }
        }

        return scales.Count == 0 ? Default.ScalePercent : scales.Average();
    }

    private static double NormalizeDisplayPercentage(double value)
    {
        return Clamp(RoundToStep(value, DisplayPercentageStep), DisplayPercentageMin, DisplayPercentageMax);
    }

    private static bool TryGetFiniteNumber(JsonNode? value, out double number)
    {
        number = 0;
        if (value is null)
        {
            return false;
        }

        try
        {
            number = value.GetValue<double>();
            return double.IsFinite(number);
        }
        catch (InvalidOperationException)
        {
            return false;
        }
        catch (FormatException)
        {
            return false;
        }
    }

    private static string JsonKey(PointerMovementSizeKey size)
    {
        return size switch
        {
            PointerMovementSizeKey.Small => "small",
            PointerMovementSizeKey.Medium => "medium",
            PointerMovementSizeKey.Large => "large",
            _ => throw new ArgumentOutOfRangeException(nameof(size), size, null)
        };
    }

    private static double RoundToStep(double value, double step)
    {
        return Math.Round(value / step, MidpointRounding.AwayFromZero) * step;
    }

    private static double Clamp(double value, double min, double max)
    {
        return Math.Min(max, Math.Max(min, value));
    }
}

public interface IPointerMovementSettingsStore
{
    PointerMovementSettings Load();

    PointerMovementSettings Save(PointerMovementSettings settings);
}

public sealed class JsonPointerMovementSettingsStore : IPointerMovementSettingsStore
{
    private readonly string filePath;
    private readonly Action<string> warn;

    public JsonPointerMovementSettingsStore(string filePath, Action<string>? warn = null)
    {
        this.filePath = filePath;
        this.warn = warn ?? Console.WriteLine;
    }

    public PointerMovementSettings Load()
    {
        try
        {
            return PointerMovementSettingsModel.Normalize(JsonNode.Parse(File.ReadAllText(filePath)));
        }
        catch (Exception error) when (error is FileNotFoundException or DirectoryNotFoundException)
        {
            return PointerMovementSettingsModel.Default;
        }
        catch
        {
            warn("Switchify pointer movement settings could not be loaded. Defaults will be used.");
            return PointerMovementSettingsModel.Default;
        }
    }

    public PointerMovementSettings Save(PointerMovementSettings settings)
    {
        PointerMovementSettings normalized = PointerMovementSettingsModel.Normalize(settings);
        string content = JsonSerializer.Serialize(PointerMovementSettingsModel.ToJsonObject(normalized), JsonOptions) + "\n";
        JsonFileStore.WriteJsonFileAtomicSync(filePath, content);
        return normalized;
    }

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        WriteIndented = true
    };
}
